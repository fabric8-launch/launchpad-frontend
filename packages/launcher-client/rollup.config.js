import peerDepsExternal from 'rollup-plugin-peer-deps-external';
import filesize from 'rollup-plugin-filesize';
import tslint from 'rollup-plugin-tslint';
import typescript from 'rollup-plugin-typescript2';
import json from 'rollup-plugin-json';
import stripCode from "rollup-plugin-strip-code"

import pkg from './package.json';
const isProduction = process.env.NODE_ENV === 'production';

const config = {
    input: 'src/index.ts',
    output: [
        {
            file: pkg.main,
            format: 'cjs',
        },
        {
            file: pkg.module,
            format: 'es',
        },
    ],
    external: [
        ...Object.keys(pkg.dependencies || {}),
        ...Object.keys(pkg.peerDependencies || {}),
    ],
    plugins: [
        peerDepsExternal(),
        tslint({
            include: './src/**/*.tsx?'
        }),
        typescript({
            clean: true,
            typescript: require('typescript'),
            verbosity: 0,
            useTsconfigDeclarationDir: true
        }),
        json(),
        isProduction && stripCode({
            start_comment: 'test-code',
            end_comment: 'end-test-code'
        }),
        filesize(),
    ],
};

export default config;
