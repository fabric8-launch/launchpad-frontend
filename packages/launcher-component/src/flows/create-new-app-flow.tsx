import * as React from 'react';
import { useState } from 'react';
import _ from 'lodash';

import { BackendForm, defaultBackendFormValue, isBackendFormValueValid, BackendFormValue, } from '../forms/backend-form';
import { defaultFrontendFormValue, FrontendForm, isFrontendFormValueValid, FrontendFormValue, } from '../forms/frontend-form';
import { BackendFormOverview } from '../forms/backend-form-overview';
import { FrontendFormOverview } from '../forms/frontend-form-overview';
import { SrcLocationForm, SrcLocationFormValue } from '../forms/src-location-form';
import { SrcLocationFormOverview } from '../forms/src-location-form-overview';
import { LaunchFlow } from './launch-flow';
import { toNewAppPayload } from './launcher-client-adapters';

interface CustomApp {
  backend: BackendFormValue;
  frontend: FrontendFormValue;
  srcLocation: SrcLocationFormValue;
}

const defaultCustomApp = {
  backend: defaultBackendFormValue,
  frontend: defaultFrontendFormValue,
  srcLocation: {
    repository: { name: 'my-app-' + _.random(1, 1000) }
  },
};

export function CreateNewAppFlow(props: { onCancel?: () => void }) {
  const [app, setApp] = useState<CustomApp>(defaultCustomApp);

  const isValidForm = () => isFrontendFormValueValid(app.frontend) && isBackendFormValueValid(app.backend);

  const items = [
    {
      id: 'frontend',
      title: 'Frontend',
      overview: {
        component: ({edit}) => (
          <FrontendFormOverview value={app.frontend} onClick={edit}/>
        ),
      },
      form: {
        component: ({close}) => (
          <FrontendForm
            value={app.frontend}
            onSave={(frontend) => {
              setApp({...app, frontend});
              close();
            }}
            onCancel={close}
          />
        ),
      }
    },
    {
      id: 'backend',
      title: 'Backend',
      overview: {
        component: ({edit}) => (
          <BackendFormOverview value={app.backend} onClick={edit}/>
        ),
      },
      form: {
        component: ({close}) => (
          <BackendForm
            value={app.backend}
            onSave={(backend) => {
              setApp({...app, backend});
              close();
            }}
            onCancel={close}
          />
        ),
      }
    },
    {
      id: 'srcLocation',
      title: 'Source Location',
      overview: {
        component: ({edit}) => (
          <SrcLocationFormOverview value={app.srcLocation} onClick={edit}/>
        ),
      },
      form: {
        component: ({close}) => (
          <SrcLocationForm
            value={app.srcLocation}
            onSave={(srcLocation) => {
              setApp({...app, srcLocation});
              close();
            }}
            onCancel={close}
          />
        ),
      }
    }
  ];

  return (
    <LaunchFlow
      items={items}
      isValid={isValidForm}
      buildAppPayload={() => toNewAppPayload(app)}
      onCancel={props.onCancel}
    />
  );

}
