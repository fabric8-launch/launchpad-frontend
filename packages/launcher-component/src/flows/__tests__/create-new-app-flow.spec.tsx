import * as React from 'react';
import 'jest-dom/extend-expect';
import { act, cleanup, fireEvent, render } from 'react-testing-library';
import { CreateNewAppFlow } from '../create-new-app-flow';
import { LauncherClientProvider } from '../..';
import { mockLauncherClient, waitForTick } from 'launcher-client';

afterEach(() => {
  console.log('cleanup()');
  cleanup();
});

jest.useFakeTimers();

async function flushPromises() {
// FIXME: use the new await version when it's available https://github.com/facebook/react/pull/14853
  act(() => {
    console.log('runAllTick()');
    jest.runAllTicks();
    jest.runOnlyPendingTimers();
    jest.runAllImmediates();
  });
  await waitForTick('act()');
}

async function configureBackend(comp, runtime, ...capabilities: string[]) {
  fireEvent.click(comp.getByLabelText('Open backend editor'));

  expect(comp.getByLabelText('Edit backend')).toBeDefined();

  // Resolve runtimes
  await flushPromises();

  fireEvent.click(comp.getByLabelText(`Choose ${runtime} as runtime`));

  // Resolve capabilities
  await flushPromises();
  // Resolve enums promises
  await flushPromises();

  capabilities.forEach(c => fireEvent.click(comp.getByLabelText(`Pick ${c} capability`)));

  fireEvent.click(comp.getByLabelText('Save backend'));

  // Resolve overview promises
  await flushPromises();
}

async function configureFrontend(comp, runtime) {
  fireEvent.click(comp.getByLabelText('Open frontend editor'));

  expect(comp.getByLabelText('Edit frontend')).toBeDefined();

  // Resolve runtimes
  await flushPromises();

  fireEvent.click(comp.getByLabelText(`Choose ${runtime} as runtime`));

  fireEvent.click(comp.getByLabelText('Save frontend'));

  // Resolve overview promises
  await flushPromises();
}

describe('<CreateNewAppFlow />', () => {
  it('renders and initializes the CreateNewAppFlow correctly', async () => {
    const comp = render(<LauncherClientProvider><CreateNewAppFlow appName="my-test-app"/></LauncherClientProvider>);
    expect(comp.getByLabelText('Loading dest-repository')).toBeDefined();
    expect(comp.getByLabelText('Loading openshift-deployment')).toBeDefined();

    // Resolve data from auto loader
    await flushPromises();

    // Resolve overview promises
    await flushPromises();

    expect(comp.getByLabelText('dest-repository is configured')).toBeDefined();
    expect(comp.getByLabelText('openshift-deployment is configured')).toBeDefined();
    expect(comp.getByLabelText('welcome-app is configured')).toBeDefined();

    expect(comp.getByLabelText('backend is not configured')).toBeDefined();
    expect(comp.getByLabelText('frontend is not configured')).toBeDefined();

    expect(comp.getByLabelText('Launch Application')).toHaveAttribute('disabled');
    expect(comp.getByLabelText('Download Application')).toHaveAttribute('disabled');
  });
  it('Configure backend and check full launch until next steps', async () => {
    const mockClient = mockLauncherClient();
    const comp = render(<LauncherClientProvider client={mockClient}><CreateNewAppFlow appName="my-test-app"/></LauncherClientProvider>);

    // Resolve data from auto loader
    await flushPromises();
    // Resolve overview promises
    await flushPromises();

    await configureBackend(comp, 'vertx', 'rest');
    expect(comp.getByLabelText('backend is configured')).toBeDefined();

    expect(comp.getByLabelText('Launch Application')).not.toHaveAttribute('disabled');
    expect(comp.getByLabelText('Download Application')).not.toHaveAttribute('disabled');

    fireEvent.click(comp.getByLabelText('Launch Application'));

    expect(comp.getByLabelText('Waiting for server response...')).toBeDefined();

    expect(mockClient.currentPayload).toMatchSnapshot('payload');

    // Resolve launch result
    await flushPromises();

    expect(comp.getByLabelText('Receiving launch progress events...')).toBeDefined();
    expect(comp.getByLabelText('GITHUB_CREATE is in-progress')).toBeDefined();
    expect(comp.getByLabelText('GITHUB_PUSHED is in-progress')).toBeDefined();
    expect(comp.getByLabelText('OPENSHIFT_CREATE is in-progress')).toBeDefined();
    expect(comp.getByLabelText('OPENSHIFT_PIPELINE is in-progress')).toBeDefined();
    expect(comp.getByLabelText('GITHUB_WEBHOOK is in-progress')).toBeDefined();

    // Resolve GITHUB_CREATE
    await flushPromises();

    expect(comp.getByLabelText('GITHUB_CREATE is completed')).toBeDefined();

    // Resolve GITHUB_PUSHED
    await flushPromises();
    expect(comp.getByLabelText('GITHUB_PUSHED is completed')).toBeDefined();

    // Resolve OPENSHIFT_CREATE
    await flushPromises();
    expect(comp.getByLabelText('OPENSHIFT_CREATE is completed')).toBeDefined();

    // Resolve OPENSHIFT_PIPELINE
    await flushPromises();
    expect(comp.getByLabelText('OPENSHIFT_PIPELINE is completed')).toBeDefined();

    // Resolve GITHUB_WEBHOOK
    await flushPromises();
    expect(comp.getByLabelText('GITHUB_WEBHOOK is completed')).toBeDefined();

    // Resolve complete progress
    await flushPromises();
    expect(comp.getByLabelText('Your Application has been launched')).toBeDefined();

    expect(comp.getByLabelText('Console link').getAttribute('href')).toMatchSnapshot('Console link');

    expect(comp.getByLabelText('Welcome Application link').getAttribute('href')).toMatchSnapshot('Welcome Application link');

    expect(comp.getByLabelText('Repository link').getAttribute('href')).toMatchSnapshot('Repository link');

  });
  it('Configure frontend, launch and check payload', async () => {
    const mockClient = mockLauncherClient();
    const comp = render(<LauncherClientProvider client={mockClient}><CreateNewAppFlow appName="my-test-app"/></LauncherClientProvider>);

    // Resolve data from auto loader
    await flushPromises();
    // Resolve overview promises
    await flushPromises();

    await configureFrontend(comp, 'react');
    expect(comp.getByLabelText('frontend is configured')).toBeDefined();

    fireEvent.click(comp.getByLabelText('Launch Application'));

    expect(mockClient.currentPayload).toMatchSnapshot('payload');

  });

  it('Configure frontend and backend, launch and check payload', async () => {
    const mockClient = mockLauncherClient();
    const comp = render(<LauncherClientProvider client={mockClient}><CreateNewAppFlow appName="my-test-app"/></LauncherClientProvider>);

    // Resolve data from auto loader
    await flushPromises();
    // Resolve overview promises
    await flushPromises();

    await configureFrontend(comp, 'react');
    expect(comp.getByLabelText('frontend is configured')).toBeDefined();

    await configureBackend(comp, 'vertx', 'rest');
    expect(comp.getByLabelText('backend is configured')).toBeDefined();

    fireEvent.click(comp.getByLabelText('Launch Application'));

    expect(mockClient.currentPayload).toMatchSnapshot('payload');
  });

  it('Configure backend with multiple capabilities, launch and check payload', async () => {
    const mockClient = mockLauncherClient();
    const comp = render(<LauncherClientProvider client={mockClient}><CreateNewAppFlow appName="my-test-app"/></LauncherClientProvider>);

    // Resolve data from auto loader
    await flushPromises();
    // Resolve overview promises
    await flushPromises();

    await configureBackend(comp, 'quarkus', 'rest', 'database');
    expect(comp.getByLabelText('backend is configured')).toBeDefined();

    fireEvent.click(comp.getByLabelText('Launch Application'));

    expect(mockClient.currentPayload).toMatchSnapshot('payload');
  });
});
