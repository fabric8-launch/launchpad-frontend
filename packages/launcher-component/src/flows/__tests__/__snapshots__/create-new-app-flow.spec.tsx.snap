// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CreateNewAppFlow /> Check that launch is working after download for the same application: payload 1`] = `
Object {
  "clusterId": "starter-ca-central-1",
  "gitOrganization": "",
  "gitRepository": "my-test-app",
  "project": Object {
    "application": "my-test-app",
    "parts": Array [
      Object {
        "capabilities": Array [
          Object {
            "module": "health",
            "props": undefined,
          },
          Object {
            "module": "rest",
            "props": undefined,
          },
          Object {
            "module": "database",
            "props": undefined,
          },
          Object {
            "module": "welcome",
          },
        ],
        "category": "backend",
        "shared": Object {
          "runtime": Object {
            "name": "quarkus",
            "version": "community",
          },
        },
      },
    ],
  },
  "projectName": "my-test-app",
}
`;

exports[`<CreateNewAppFlow /> Configure backend and check full download until next steps: Download link 1`] = `"http://mock/result.zip"`;

exports[`<CreateNewAppFlow /> Configure backend and check full download until next steps: payload 1`] = `
Object {
  "clusterId": "starter-ca-central-1",
  "gitOrganization": "",
  "gitRepository": "my-test-app",
  "project": Object {
    "application": "my-test-app",
    "parts": Array [
      Object {
        "capabilities": Array [
          Object {
            "module": "health",
            "props": undefined,
          },
          Object {
            "module": "rest",
            "props": undefined,
          },
          Object {
            "module": "database",
            "props": undefined,
          },
          Object {
            "module": "welcome",
          },
        ],
        "category": "backend",
        "shared": Object {
          "runtime": Object {
            "name": "quarkus",
            "version": "community",
          },
        },
      },
    ],
  },
  "projectName": "my-test-app",
}
`;

exports[`<CreateNewAppFlow /> Configure backend and check full launch until next steps: Console link 1`] = `"https://console.starter-us-east-2.openshift.com/console/projects"`;

exports[`<CreateNewAppFlow /> Configure backend and check full launch until next steps: Repository link 1`] = `"https://github.com/fabric8-launcher/launcher-backend"`;

exports[`<CreateNewAppFlow /> Configure backend and check full launch until next steps: Welcome Application link 1`] = `"http://welcome-gullible-rake.7e14.starter-us-west-2.openshiftapps.com/"`;

exports[`<CreateNewAppFlow /> Configure backend and check full launch until next steps: payload 1`] = `
Object {
  "clusterId": "starter-ca-central-1",
  "gitOrganization": "",
  "gitRepository": "my-test-app",
  "project": Object {
    "application": "my-test-app",
    "parts": Array [
      Object {
        "capabilities": Array [
          Object {
            "module": "health",
            "props": undefined,
          },
          Object {
            "module": "rest",
            "props": undefined,
          },
          Object {
            "module": "welcome",
          },
        ],
        "category": "backend",
        "shared": Object {
          "runtime": Object {
            "name": "vertx",
            "version": "community",
          },
        },
      },
    ],
  },
  "projectName": "my-test-app",
}
`;

exports[`<CreateNewAppFlow /> Configure backend with multiple capabilities, launch and check payload: payload 1`] = `
Object {
  "clusterId": "starter-ca-central-1",
  "gitOrganization": "",
  "gitRepository": "my-test-app",
  "project": Object {
    "application": "my-test-app",
    "parts": Array [
      Object {
        "capabilities": Array [
          Object {
            "module": "health",
            "props": undefined,
          },
          Object {
            "module": "rest",
            "props": undefined,
          },
          Object {
            "module": "database",
            "props": undefined,
          },
          Object {
            "module": "welcome",
          },
        ],
        "category": "backend",
        "shared": Object {
          "runtime": Object {
            "name": "quarkus",
            "version": "community",
          },
        },
      },
    ],
  },
  "projectName": "my-test-app",
}
`;

exports[`<CreateNewAppFlow /> Configure backend with no capability, launch and check payload: payload 1`] = `
Object {
  "clusterId": "starter-ca-central-1",
  "gitOrganization": "",
  "gitRepository": "my-test-app",
  "project": Object {
    "application": "my-test-app",
    "parts": Array [
      Object {
        "capabilities": Array [
          Object {
            "module": "health",
            "props": undefined,
          },
          Object {
            "module": "welcome",
          },
        ],
        "category": "backend",
        "shared": Object {
          "runtime": Object {
            "name": "quarkus",
            "version": "community",
          },
        },
      },
    ],
  },
  "projectName": "my-test-app",
}
`;

exports[`<CreateNewAppFlow /> Configure frontend and backend, launch and check payload: payload 1`] = `
Object {
  "clusterId": "starter-ca-central-1",
  "gitOrganization": "",
  "gitRepository": "my-test-app",
  "project": Object {
    "application": "my-test-app",
    "parts": Array [
      Object {
        "capabilities": Array [
          Object {
            "module": "web-app",
          },
        ],
        "category": "frontend",
        "shared": Object {
          "runtime": Object {
            "name": "react",
            "version": "community",
          },
        },
        "subFolderName": "frontend",
      },
      Object {
        "capabilities": Array [
          Object {
            "module": "health",
            "props": undefined,
          },
          Object {
            "module": "rest",
            "props": undefined,
          },
        ],
        "category": "backend",
        "shared": Object {
          "runtime": Object {
            "name": "vertx",
            "version": "community",
          },
        },
        "subFolderName": "backend",
      },
      Object {
        "capabilities": Array [
          Object {
            "module": "welcome",
          },
        ],
        "category": "support",
        "subFolderName": "support",
      },
    ],
  },
  "projectName": "my-test-app",
}
`;

exports[`<CreateNewAppFlow /> Configure frontend, launch and check payload: payload 1`] = `
Object {
  "clusterId": "starter-ca-central-1",
  "gitOrganization": "",
  "gitRepository": "my-test-app",
  "project": Object {
    "application": "my-test-app",
    "parts": Array [
      Object {
        "capabilities": Array [
          Object {
            "module": "web-app",
          },
          Object {
            "module": "welcome",
          },
        ],
        "category": "frontend",
        "shared": Object {
          "runtime": Object {
            "name": "react",
            "version": "community",
          },
        },
      },
    ],
  },
  "projectName": "my-test-app",
}
`;
